internal/irzstd/writer.go

```diff
@@ -0,0 +1,112 @@
+// Package implements writer that converts log events to Zstd compressed IR. Effectively chaining
+// together [ir.Writer] and [zstd.Encoder] in series.
+
+package irzstd
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/klauspost/compress/zstd"
+
+	"github.com/y-scope/clp-ffi-go/ffi"
+	"github.com/y-scope/clp-ffi-go/ir"
+)
+
+type Writer interface {
+	// Converts log events to Zstd compressed IR and outputs to the Zstd buffer.
+	//
+	// Parameters:
+	//   - logEvents: A slice of log events to be encoded
+	//
+	// Returns:
+	//   - err
+	WriteIrZstd([]ffi.LogEvent) error
+
+	// Closes IR stream and Zstd frame. After calling close, Writer must be Reset() prior to calling
+	// write.
+	//
+	// Returns:
+	//   - err
+	CloseStreams() error
+
+	// Closes Writer.
+	//
+	// Returns:
+	//   - err
+	Close() error
+
+	// Reinitialize Writer after calling CloseStreams().
+	//
+	// Returns:
+	//   - err
+	Reset() error
+
+	// Getter for useDiskBuffer.
+	//
+	// Returns:
+	//   - useDiskBuffer: On/off for disk buffering
+	GetUseDiskBuffer() bool
+
+	// Getter for Zstd Output.
+	//
+	// Returns:
+	//   - zstdOutput: Reader for Zstd output
+	GetZstdOutput() io.Reader
+
+	// Get size of Zstd output.
+	//
+	// Returns:
+	//	 - size: Bytes written
+	//   - err
+	GetZstdOutputSize() (int, error)
+}
+
+// Writes log events to a IR Writer.
+//
+// Parameters:
+//   - irWriter: CLP IR writer to write each log event with
+//   - logEvents: A slice of log events to be encoded
+//
+// Returns:
+//   - err: error if an event could not be written
+func writeIr(irWriter *ir.Writer, logEvents []ffi.LogEvent) error {
+	for _, event := range logEvents {
+		_, err := irWriter.Write(event)
+		if err != nil {
+			err = fmt.Errorf("failed to encode event %v into ir: %w", event, err)
+			return err
+		}
+	}
+	return nil
+}
```

Comment:
We should somehow inform/propagate the number of log events we successfully write, as it will be important in the case of a failure.
Just counting and returning the number from here and then `WriteIrZstd` works fine with little change. Another option could be to have `WriteIrZstd` only take a single log event (would require more refactoring and puts the loop outside the writer code). Any solution should be fine.